### PR TITLE
Store Validators Min and Max Spans for Slashing

### PR DESCRIPTION
**Pull request by shayzluf**
[Part of] #3133

---

# Description

Store min max span per epoch per validator in order to support https://github.com/protolambda/eth2-surround/blob/master/README.md#min-max-surround

### CODE REVIEW

File: proto/eth/v1alpha1/slasher.proto [hunk 1]
```diff
@@ -47,3 +47,12 @@ message ProposerSlashingRequest {
 message ProposerSlashingResponse {
     repeated ethereum.eth.v1alpha1.ProposerSlashing proposer_slashing=1;
 }
+
+message MinMaxSpan {
+    uint32 min_span=1;
+    uint32 max_span=2;
+}
+
+message EpochSpanMap{
```
Comment (terencechain) [inline proto/eth/v1alpha1/slasher.proto]: Please add descriptions on what this map is doing, especially what the key is doing

File: proto/eth/v1alpha1/slasher.proto [hunk 2]
```diff
@@ -47,3 +47,12 @@ message ProposerSlashingRequest {
 message ProposerSlashingResponse {
     repeated ethereum.eth.v1alpha1.ProposerSlashing proposer_slashing=1;
 }
+
+message MinMaxSpan {
+    uint32 min_span=1;
```
Comment (terencechain) [inline proto/eth/v1alpha1/slasher.proto]: Please add descriptions on what MinMaxSpan is

File: slasher/db/min_max-span_test.go [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+
+func TestNilDBValidatorSpanMap(t *testing.T) {
+	db := SetupSlasherDB(t)
+	defer TeardownSlasherDB(t, db)
+
+	validatorID := uint64(1)
+
+	vsm, err := db.ValidatorSpansMap(validatorID)
+	if err != nil {
+		t.Fatal("nil ValidatorSpansMap should not return error")
```
Comment (terencechain) [inline slasher/db/min_max-span_test.go]: Maybe log `err` too for completeness

File: slasher/db/min_max-span_test.go [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+			t.Fatalf("save validator span map failed: %v", err)
+		}
+
+		sm, err := db.ValidatorSpansMap(tt.validatorID)
+		if err != nil {
+			t.Fatalf("failed to get validator span map: %v", err)
+		}
+
+		if sm == nil || !proto.Equal(sm, tt.spanMap) {
+			t.Fatalf("get should return validator span map: %v", tt.spanMap)
```
Comment (terencechain) [inline slasher/db/min_max-span_test.go]: Add `sm` to log too

File: slasher/db/min_max-span_test.go [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+			t.Fatalf("delete validator span map error: %v", err)
+		}
+		sm, err = db.ValidatorSpansMap(tt.validatorID)
+		if err != nil {
+			t.Fatal(err)
+		}
+		if sm.EpochSpanMap != nil {
+			t.Errorf("Expected validator span map to be deleted, received: %v", sm)
+		}
+
```
Comment (terencechain) [inline slasher/db/min_max-span_test.go]:
```suggestion
```

File: slasher/db/min_max-span_test.go [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+	}
+
```
Comment (terencechain) [inline slasher/db/min_max-span_test.go]:
```suggestion
```

File: proto/eth/v1alpha1/slasher.proto (hunk 2 above)
Comment (shayzluf) [reply proto/eth/v1alpha1/slasher.proto]: added. thanks

File: proto/eth/v1alpha1/slasher.proto (hunk 1 above)
Comment (shayzluf) [reply proto/eth/v1alpha1/slasher.proto]: added. thanks

File: slasher/db/min_max-span_test.go (hunk 3 above)
Comment (shayzluf) [reply slasher/db/min_max-span_test.go]: added

File: proto/eth/v1alpha1/slasher.proto (hunk 2 above)
Comment (rauljordan) [reply proto/eth/v1alpha1/slasher.proto]: Also formatting in this file should be of the form `uint32 min_span = 1`; Maybe this proto can have better naming? `MinMaxSpan` doesn't convey much info to the reader or someone who is unfamiliar with the algorithm. I guess it should be fine it the link to the protolambda README is provided, but just thinking if there are better names...

File: proto/eth/v1alpha1/slasher.proto [hunk 7]
```diff
@@ -47,3 +47,17 @@ message ProposerSlashingRequest {
...
+// each attestation source to those spans
+// see https://github.com/protolambda/eth2-surround/blob/master/README.md#min-max-surround
+// for further details.
+message MinMaxSpan {
+    uint32 min_span=1;
+    uint32 max_span=2;
+}
+// every validator will have his own spans map containing min and max value for each epoch
+// in order to detect slashable attestation as quick as possible.
+message EpochSpanMap{
```
Comment (rauljordan) [inline proto/eth/v1alpha1/slasher.proto]:
```suggestion
message EpochSpanMap {
```

File: slasher/db/min_max-span_test.go [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+type spanMapTestStruct struct {
+	validatorID uint64
+	spanMap     *ethpb.EpochSpanMap
+}
+
+var spanTests []spanMapTestStruct
+
+func init() {
+	spanTests = []spanMapTestStruct{
```
Comment (rauljordan) [inline slasher/db/min_max-span_test.go]: Maybe each of these can have better descriptions as to what exactly they are or maybe why the values were selected this way

File: slasher/db/min_max-span_test.go [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+
+func TestDeleteValidatorSpanMap(t *testing.T) {
+	db := SetupSlasherDB(t)
+	defer TeardownSlasherDB(t, db)
+
+	for _, tt := range spanTests {
+
+		err := db.SaveValidatorSpansMap(tt.validatorID, tt.spanMap)
+		if err != nil {
+			t.Fatalf("save validator span map failed: %v", err)
```
Comment (rauljordan) [inline slasher/db/min_max-span_test.go]:
```suggestion
			t.Fatalf("Save validator span map failed: %v", err)
```
in general formatting in t.Error or t.Fatal calls is capitalized based on the rest of our repos

File: slasher/db/min_max_span.go [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+	return db.update(func(tx *bolt.Tx) error {
+		bucket := tx.Bucket(validatorsMinMaxSpanBucket)
+		key := bytesutil.Bytes4(validatorID)
+		enc := bucket.Get(key)
+		if enc == nil {
+			return nil
+		}
+		if err := bucket.Delete(key); err != nil {
+			tx.Rollback()
+			return errors.Wrapf(err, "failed to delete the span map for validator id: %v from validators min max span bucket", validatorID)
```
Comment (rauljordan) [inline slasher/db/min_max_span.go]:
```suggestion
			return errors.Wrapf(err, "failed to delete validator id: %d from validators min max span bucket", validatorID)
```

File: slasher/db/min_max_span.go [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+		}
+		if err := bucket.Put(key, val); err != nil {
+			return errors.Wrapf(err, "failed to include the span map for validator id: %v in the validators min max span bucket", validatorID)
+		}
+		return err
+	})
+	return err
+}
+
+// DeleteValidatorSpanMap deletes a validator span map using the validator id as bucket key.
```
Comment (rauljordan) [inline slasher/db/min_max_span.go]: Are these validator indices or ID's? What's the difference? Can we stick to calling them validator indices? (idx)

File: slasher/db/min_max_span.go [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+func (db *Store) SaveValidatorSpansMap(validatorID uint64, spanMap *ethpb.EpochSpanMap) error {
+	err := db.update(func(tx *bolt.Tx) error {
+		bucket := tx.Bucket(validatorsMinMaxSpanBucket)
+		key := bytesutil.Bytes4(validatorID)
+		val, err := proto.Marshal(spanMap)
+		if err != nil {
+			return errors.Wrap(err, "failed to marshal span map")
+		}
+		if err := bucket.Put(key, val); err != nil {
+			return errors.Wrapf(err, "failed to include the span map for validator id: %v in the validators min max span bucket", validatorID)
```
Comment (rauljordan) [inline slasher/db/min_max_span.go]: Line too long, see suggestion below

File: slasher/db/min_max-span_test.go [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+
+import (
+	"github.com/gogo/protobuf/proto"
+	"testing"
+
+	ethpb "github.com/prysmaticlabs/prysm/proto/eth/v1alpha1"
+)
+
+type spanMapTestStruct struct {
+	validatorID uint64
```
Comment (rauljordan) [inline slasher/db/min_max-span_test.go]: Is this the index or an ID? Can we call validatorIdx if it is the index in the state.Validator registry?

File: slasher/db/min_max-span_test.go [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+	vsm, err := db.ValidatorSpansMap(validatorID)
+	if err != nil {
+		t.Fatalf("nil ValidatorSpansMap should not return error: %v", err)
+	}
+	if vsm.EpochSpanMap != nil {
+		t.Fatal("ValidatorSpansMap should return nil")
+	}
+}
+
+func TestSaveValidatorSpanMap(t *testing.T) {
```
Comment (rauljordan) [inline slasher/db/min_max-span_test.go]: Would recommend naming these tests with the form
```
func TestValidatorSpanMap_Save
func TestValidatorSpanMap_Delete
func TestValidatorSpanMap_NilDB
```
for better go and bazel test filtering rules

```

File: slasher/db/schema.go [hunk 15]
```diff
@@ -10,6 +10,7 @@ var (
 	historicBlockHeadersBucket        = []byte("historic-block-headers-bucket")
 	indexedAttestationsIndicesBucket  = []byte("indexed-attestations-indices-bucket")
 	validatorsPublicKeysBucket        = []byte("validators-public-keys-bucket")
+	validatorsMinMaxSpanBucket        = []byte("validators-min-max-span-bucket")
```
Comment (rauljordan) [inline slasher/db/schema.go]: Can we add a comment on what this is used for? Once again, for an inexperienced or new contributor, the name behind this doesn't convey much context and it is not clear what it is needed for.

Comment (shayzluf) [reply slasher/db/schema.go]: sure. adding

File: slasher/db/min_max-span_test.go (hunk 8 above)
Comment (shayzluf) [reply slasher/db/min_max-span_test.go]: i don't have the detection and update phase in place. so those are just examples. will be adding detection and logic in the next pr

File: slasher/db/min_max-span_test.go (hunk 9 above)
Comment (shayzluf) [reply slasher/db/min_max-span_test.go]: thanks

Comment (shayzluf) [reply slasher/db/min_max-span_test.go]: fixed

File: slasher/db/min_max_span.go (hunk 10 above)
Comment (shayzluf) [reply slasher/db/min_max_span.go]: thanks

File: slasher/db/min_max-span_test.go (hunk 14 above)
Comment (shayzluf) [reply slasher/db/min_max-span_test.go]: done

File: slasher/db/min_max-span_test.go (hunk 13 above)
Comment (shayzluf) [reply slasher/db/min_max-span_test.go]: changed all

File: slasher/db/min_max_span.go (hunk 12 above)
Comment (shayzluf) [reply slasher/db/min_max_span.go]: changed. thanks

File: slasher/db/min_max_span.go (hunk 11 above)
Comment (shayzluf) [reply slasher/db/min_max_span.go]: changed

File: proto/eth/v1alpha1/slasher.proto (hunk 7 above)
Comment (shayzluf) [reply proto/eth/v1alpha1/slasher.proto]: fixed

File: proto/eth/v1alpha1/slasher.proto [hunk 16]
```diff
@@ -41,9 +41,23 @@ message ValidatorIDToIdxAttList {
...
-    ethereum.eth.v1alpha1.BeaconBlockHeader block_header=1;
-    uint64 validator_index=2;
+    ethereum.eth.v1alpha1.BeaconBlockHeader block_header = 1;
+    uint64 validator_index = 2;
 }
 message ProposerSlashingResponse {
-    repeated ethereum.eth.v1alpha1.ProposerSlashing proposer_slashing=1;
+    repeated ethereum.eth.v1alpha1.ProposerSlashing proposer_slashing = 1;
+}
+// In order to detect surround or surrounded attestation we need to compare
```
Comment (terencechain) [inline proto/eth/v1alpha1/slasher.proto]:
```suggestion
// In order to detect surrounded attestation we need to compare
```

File: slasher/db/min_max-span_test.go [hunk 17]
```diff
@@ -0,0 +1,118 @@
...
+		}
+	}
+
+	for _, tt := range spanTests {
+		sm, err := db.ValidatorSpansMap(tt.validatorIdx)
+		if err != nil {
+			t.Fatalf("Failed to get validator span map: %v", err)
+		}
+		if sm == nil || !proto.Equal(sm, tt.spanMap) {
+			t.Fatalf("Get should return validator span map: %v", sm)
```
Comment (terencechain) [inline slasher/db/min_max-span_test.go]:
```suggestion
			t.Fatalf("Get should return validator span map: %v got: %v", tt.spanMap, sm)
```

File: slasher/db/min_max_span.go [hunk 18]
```diff
@@ -0,0 +1,70 @@
...
+func createEpochSpanMap(enc []byte) (*ethpb.EpochSpanMap, error) {
+	epochSpanMap := &ethpb.EpochSpanMap{}
+	err := proto.Unmarshal(enc, epochSpanMap)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to unmarshal encoding")
+	}
+	return epochSpanMap, nil
+}
+
+// ValidatorSpansMap accepts validator indices and returns the corresponding spans
```
Comment (terencechain) [inline slasher/db/min_max_span.go]:
```suggestion
// ValidatorSpansMap accepts validator index and returns the corresponding spans
```

File: slasher/db/min_max_span.go [hunk 19]
```diff
@@ -0,0 +1,70 @@
...
+// map for slashing detection and update.
```
Comment (terencechain) [inline slasher/db/min_max_span.go]:
```suggestion
// map for slashing detection.
```

File: slasher/db/min_max_span.go [hunk 20]
```diff
@@ -0,0 +1,70 @@
...
+// Returns nil if the span map for this validator indices does not exist.
```
Comment (terencechain) [inline slasher/db/min_max_span.go]:
```suggestion
// Returns nil if the span map for this validator index does not exist.
```

File: slasher/db/min_max_span.go [hunk 21]
```diff
@@ -0,0 +1,70 @@
...
+		sm, err = createEpochSpanMap(enc)
+		if err != nil {
+			return err
+		}
+		return nil
+	})
+	return sm, err
+}
+
+// SaveValidatorSpansMap accepts validator indices and span map and writes it to disk.
```
Comment (terencechain) [inline slasher/db/min_max_span.go]:
```suggestion
// SaveValidatorSpansMap accepts a validator index and span map and writes it to disk.
```

File: slasher/db/min_max_span.go [hunk 22]
```diff
@@ -0,0 +1,70 @@
...
+		}
+		if err := bucket.Put(key, val); err != nil {
+			return errors.Wrapf(err, "failed to delete validator id: %d from validators min max span bucket", validatorIdx)
+		}
+		return err
+	})
+	return err
+}
+
+// DeleteValidatorSpanMap deletes a validator span map using the validator indices as bucket key.
```
Comment (terencechain) [inline slasher/db/min_max_span.go]:
```suggestion
// DeleteValidatorSpanMap deletes a validator span map using a validator index as bucket key.
```

**Comment by shayzluf** [comment]
@rauljordan updated the formatting

File: slasher/db/min_max-span_test.go [hunk 23]
```diff
@@ -0,0 +1,118 @@
...
+		}
+	}
+}
+
+func TestValidatorSpanMap_Delete(t *testing.T) {
+	db := SetupSlasherDB(t)
+	defer TeardownSlasherDB(t, db)
+
+	for _, tt := range spanTests {
+
```
Comment (rauljordan) [inline slasher/db/min_max-span_test.go]:
```suggestion
```

**Comment by rauljordan** [comment]
Good start on this, looking forward to seeing it materialize even more

**Pull request merged**